Library/Homebrew/software_spec.rb

```diff
@@ -513,7 +520,30 @@ def compatible_locations?(tag: Utils::Bottles.tag)
       tag.default_cellar
     end
 
-    return true if [:any, :any_skip_relocation].include?(cellar)
+    return true if RELOCATABLE_CELLARS.include?(cellar)
+
+    prefix = Pathname(cellar).parent.to_s
+
+    cellar_relocatable = cellar.size >= HOMEBREW_CELLAR.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
+    prefix_relocatable = prefix.size >= HOMEBREW_PREFIX.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
+
+    compatible_cellar = cellar == HOMEBREW_CELLAR.to_s || cellar_relocatable
+    compatible_prefix = prefix == HOMEBREW_PREFIX.to_s || prefix_relocatable
+
+    compatible_cellar && compatible_prefix
+  end
+
+  # Should the build prefix for the {Bottle} this {BottleSpecification} belongs to be relocated?
+  sig { params(tag: Utils::Bottles::Tag).returns(T::Boolean) }
+  def skip_prefix_relocation?(tag: Utils::Bottles.tag)
```

Comment:
This method seems to have a lot of overlap with `#compatible_locations?`. I wonder if they really need to be separate methods.

Reply:
Agreed.

Reply:
I had to think the about this a bit but I think the proper solution here is to have a `cellar_from_tag` method which returns the cellar.  Then both times where we need to know the bottle cellar, we can just get it from the tag.